components/push/ffi/src/lib.rs

```diff
@@ -0,0 +1,224 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use ffi_support::{
+    define_bytebuffer_destructor, define_handle_map_deleter, define_string_destructor,
+    ConcurrentHandleMap, ExternError,
+};
+use std::os::raw::c_char;
+// use sync15::telemetry;
+
+use base64;
+use lazy_static;
+use serde_json::{self, json};
+
+use config::PushConfiguration;
+use crypto::{self, Cryptography};
+use push_errors::{self, Result};
+use storage::Storage;
+use subscriber::PushManager;
+
+#[no_mangle]
+pub extern "C" fn push_enable_logcat_logging() {
+    #[cfg(target_os = "android")]
+    {
+        let _ = std::panic::catch_unwind(|| {
+            android_logger::init_once(
+                android_logger::Filter::default().with_min_level(log::Level::Debug),
+                Some("libpush_ffi"),
+            );
+            log::debug!("Android logging should be hooked up!")
+        });
+    }
+}
+
+lazy_static::lazy_static! {
+    static ref MANAGER: ConcurrentHandleMap<PushManager> = ConcurrentHandleMap::new();
+}
+
+/// Instantiate a Http connection. Returned connection must be freed with
+/// `push_connection_destroy`. Returns null and logs on errors (for now).
+#[no_mangle]
+pub unsafe extern "C" fn push_connection_new(
+    server_host: *const c_char,
+    socket_protocol: *const c_char,
+    bridge_type: *const c_char,
+    registration_id: *const c_char,
+    sender_id: *const c_char,
+    database_path: *const c_char,
+    error: &mut ExternError,
+) -> u64 {
+    log::debug!(
+        "push_connection_new {:?} {:?} -> {:?} {:?}=>{:?}",
+        socket_protocol,
+        server_host,
+        bridge_type,
+        sender_id,
+        registration_id
+    );
+    // return this as a reference to the map since that map contains the actual handles that rust uses.
```

Comment:
All of this needs to happen inside a call_with_result block (which `insert_with_result` contains), these functions can panic, which is UB if it happens over the FFI